crates/holochain/src/core/workflow/call_zome_workflow.rs

```diff
@@ -189,22 +182,34 @@ where
     ) {
         let scratch_records = workspace.source_chain().scratch_records()?;
         if scratch_records.len() == 1 {
-            let lock = holochain_state::source_chain::lock_for_entry(
+            let lock_subject = holochain_state::source_chain::chain_lock_subject_for_entry(
                 scratch_records[0].entry().as_option(),
             )?;
-            if !lock.is_empty()
-                && workspace
-                    .source_chain()
-                    .is_chain_locked(Vec::with_capacity(0))
-                    .await?
-                && !workspace.source_chain().is_chain_locked(lock).await?
-            {
-                if let Err(error) = workspace.source_chain().unlock_chain().await {
-                    tracing::error!(?error);
+
+            // If this wasn't a countersigning commit then the lock will be empty.
+            if !lock_subject.is_empty() {
+                // Otherwise, we can check whether the chain was locked with a subject matching
+                // the entry that failed validation.
+                if let Some(chain_lock) = workspace.source_chain().get_chain_lock().await? {
+                    // Here we know the chain is locked, and if the lock subject matches the entry
+                    // that the app was trying to commit then we can unlock the chain.
+                    if chain_lock.subject() == lock_subject {
+                        if let Err(error) = workspace.source_chain().unlock_chain().await {
+                            tracing::error!(?error);
+                        }
+
+                        // Immediately unlocking the chain is safe because we know that the
+                        // countersigning commit hasn't been written to the source chain here.
+                        // We still to clean up the session state in the countersigning workspace
+                        // though, so trigger the countersigning workflow and let it figure out
+                        // what happened.
+                        trigger_countersigning.trigger(&"invalid_countersigning_commit");
```

Comment:
- `&&str` since it's not a fat pointer?
- if conductor restarted here before the trigger was processed, would everything work out even though the workflow wouldn't get this `"invalid_countersigning_commit"` param? Or is it not a param, just a debug note ore something?

Reply:
- I don't know why somebody decided that this takes an `&&str` as an input. It's not a parameter, just information about where the signal came from. But the receiver doesn't actually get it. It's another one of those half-way thought through features... Just occasionally it's useful to add a debug statement in the queue consumer to see what triggers are coming through, and maybe that's what it's for? I don't really know! It would actually be quite useful in this workflow to know which is the init signal, because we shouldn't need to refresh on any other signal.
- We don't have to wonder, added a new test `chain_unlocked_outside_workflow_then_restart`. Because the entry didn't get written, the only state was the chain lock and the workspace state (in memory). So having done the unlock and the restart will have cleared the in memory state, there's nothing to do. The user won't get a signal but I think that's fine.